### DISCUSSION TITLE
Binding values to nested dictionary used with a table

### DISCUSSION THREAD

**Discussion by AnonyDon** [zauberzeug/nicegui#3455] [Q&A] [answered]
### Question

Hello! I've tried quite a few approaches to get this to work but haven't had any luck...
Please forgive any obvious oversights!

I'm trying to get the checkboxes to bind with the values of the last row clicked on's 'conditions' dictionary...
```python
from nicegui import ui, events

def change_picture(e: events.GenericEventArguments):
    image.source = e.args[1]["img"]
    global last_image_id, last_image_name
    last_image_id = e.args[1]['id']
    last_image_name = e.args[1]['name']

checkbox_dict = {}
conditions_dict = {
    'blinded': False,
    'charmed': False,
    'dazed': False,
    'deafened': False,
    'exhaustion': False,
    'frightened': False,
    'grappled': False,
    'incapacitated': False,
    'invisible': False,
    'paralyzed': False,
    'petrified': False,
    'poisoned': False,
    'prone': False,
    'restrained': False,
    'stunned': False,
    'unconscious': False,
}
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name', 'required': True, 'align': 'left'},
    {'name': 'ac', 'label': 'AC', 'field': 'ac', 'align': 'left'},
    {'name': 'hp', 'label': 'HP Current', 'field': 'hp', 'align': 'left'},
    {'name': 'type', 'label': 'Type', 'field': 'type', 'align': 'left'},
]
rows = [
    {
        'id': 0, 'name': 'Enemy', 'ac': 12,
        'type': 'enemy', 'hp': 50, 'conditions': conditions_dict,
        'img': 'images/goblin.png',
    },
    {
        'id': 1, 'name': 'Enemy 2', 'ac': 12,
        'type': 'enemy', 'hp': 50, 'conditions': conditions_dict,
        'img': 'images/goblin.png',
    },
    {
        'id': 2, 'name': 'Boss', 'ac': 12,
        'type': 'enemy', 'hp': 50, 'conditions': conditions_dict,
        'img': 'images/bargnot.png',
    },
    {
        'id': 3, 'name': 'Minion', 'ac': 12,
        'type': 'minion', 'hp': 50, 'conditions': conditions_dict,
        'img': 'images/minion.png',
    },
]

last_image_id = 0
last_image_name = "Enemy"

with ui.row(wrap=False).classes('w-full'):
    table = ui.table(columns=columns, rows=rows, selection='multiple'
                     ).props("dense separator=cell"
                             ).on('rowClick', change_picture)
    with ui.card():
        ui.label('Conditions of:')
        ui.label('name').bind_text_from(globals(), "last_image_name")
        ui.label('id').bind_text_from(globals(), "last_image_id")
        for item in conditions_dict:
            with ui.checkbox(text=f'{item}'
                             ).props('dense').bind_value(rows[last_image_id]['conditions'], item
                                                         ) as checkbox_dict[item]:
                pass
    image = ui.image('images/goblin.png')

ui.run()
```
I specifically cannot use the table's selection due to having that tied up in a bunch of other functions in what I'm making...

**Comment by falkoschindler** [comment, marked as answer]
Hi @AnonyDon,

I started experimenting with your code, but noticed that it's still a bit unclear what you're trying to do:

- Do the images have anything to do with your question? If not, please remove them from your code example. In general, try to [minimize examples](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example) as much as possible.
- What exactly does not work? What do you expect to happen? What happens instead?

If I guess correctly, you want to toggle "condition" checkboxes and they should restore their state when clicking back and forth between rows. I think there are two problems:

1. Each row contains a reference to the same dictionary instance. So changing the state in one of them changes it for all rows.
2. Currently you're binding to the first dictionary. But you want to switch between multiple dictionaries, which is probably not possible with a simple binding.

I'd recommend to simply create multiple cards, each binding to a single dictionary, and switching visibility with another binding:
```py
selected = {'id': None}

conditions = {
    'blinded': False,
    'charmed': False,
    'dazed': False,
}
columns = [{'name': 'name', 'label': 'Name', 'field': 'name', 'align': 'left'}]
rows = [
    {'id': 0, 'name': 'Enemy', 'conditions': {**conditions}},
    {'id': 1, 'name': 'Boss', 'conditions': {**conditions}},
    {'id': 2, 'name': 'Minion', 'conditions': {**conditions}},
]

ui.table(columns=columns, rows=rows).on('rowClick', lambda e: selected.update({'id': e.args[1]['id']}))
for row in rows:
    with ui.card().bind_visibility_from(selected, 'id', value=row['id']):
        ui.label(row['name'])
        for key in row['conditions']:
            ui.checkbox(key).bind_value(row['conditions'], key)
```
Sure, this creates more UI elements that need to be sent to the client. But for reasonable small tables this shouldn't be an issue. Alternatively you could use a `@ui.refreshable` to dynamically recreate the card whenever the selected row changes.

**Reply by AnonyDon** [reply]
Hey @falkoschindler, thanks again for your help, I've learned a lot from you recently.
Please forgive the lack of clarity, I'll endeavor to do better moving forward!

Your guess was spot on, and it seems what I was trying to do with binding values isn't possible, which is the conclusion I reached initially but wanted to make sure.

The images are only partially related, as they are displayed using the table's .on('rowClick'...) and I was trying to have a card with 16 different checkboxes that would be bound with the conditions values on the same row associated with the currently displayed image.

In retrospect I was a bit lazy when making the example, I'm sorry.

This is all part of an application to track combat in a TTRPG, which is probably suffering from feature creep at this point haha.
Generally the table on it will have 10-20 rows so I'm not sure if making over 300 checkboxes is a good idea. (I'm really not sure please tell me!)

The table has selection='multiple' since there are other functions that involve updating HP values of multiple rows simultaneously, so I figured that would make any kind of conditions display incompatible with the table's selection feature.

The example you provided is really interesting! Having the selected variable be a dict makes so much sense. 
Two questions though:
1. What is the difference between assigning the row's conditions using **kwargs instead of just the conditions dict itself?
2. What is the significance of the "e:" in the .on('rowClick'...) method?

Wow, the ui.refreshable decorator looks perfect! I appreciate you!

**Reply by falkoschindler** [reply]
No worries, sometimes I'm a bit impatient when looking into code examples. But in the end your intention wasn't too hard to understand. It's a great example for handling table data, user events, UI updates, binding... 🙂

- 300 checkboxes shouldn't be too much of an issue. But if in doubt, try it out and see if latency is an issue. Often it's better to keep it simple first and optimize later - if you really need to.
- With `{**conditions}` I'm creating a new dictionary `{...}` and use the keyword expansion `**` to fill it with the content of `conditions`. This way each row gets its own dictionary instance. If you simply assign the same dictionary to each row, modifying a value in one of them will modify it for all rows simultaneously, because all dictionary are the very same object.
    Oh, I just noticed that you can also write `dict(conditions)`. And you could also use `copy()` from Pythons "copy" module.
- `e` is a common name for the event arguments passed to an event handler. In this case it is an instance of the `events.GenericEventArguments` and contains the `sender` element of the event, its `client` as well as the `args` object containing the actual event data.

**Reply by AnonyDon** [reply]
Understandable. I know how frustrating it can be when you are trying to help someone and they aren't giving you the info you need.
Thanks to you, I got what I was going for working perfectly with:
```python
@ui.refreshable
def checkbox_card() -> None:
    with ui.card():
        ui.label(rows[selected['id']]['name'])
        for key in rows[selected['id']]['conditions']:
            ui.checkbox(key).bind_value(rows[selected['id']]['conditions'], key).props('dense')
```
